processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -31,25 +38,180 @@
+    }
+
+    @Override
+    public byte[] toBytes(String val)
+    {
+      if (val == null) {
+        return ByteArrays.EMPTY_ARRAY;
+      }
+      return StringUtils.toUtf8(val);
+    }
+
+    @Override
+    public int compare(String o1, String o2)
+    {
+      return ORDERING.compare(o1, o2);
+    }
+  };
 
-  private int indexOffset;
+  private final ObjectStrategy<T> strategy;
+  private final boolean allowReverseLookup;
+  private final int size;
+  private final BufferIndexed bufferIndexed;
+  private final List<ByteBuffer> valueBuffers;
+  private final ByteBuffer headerBuffer;
+  private final int numElementsPerValueFile;
+
+  private volatile ByteBuffer theBuffer;
+
+  GenericIndexed(
+      ByteBuffer buffer,
+      ObjectStrategy<T> strategy,
+      boolean allowReverseLookup
+  )
+  {
+    this.theBuffer = buffer;
+    this.strategy = strategy;
+    this.allowReverseLookup = allowReverseLookup;
+    size = theBuffer.getInt();
+    numElementsPerValueFile = size;
+
+    int indexOffset = theBuffer.position();
+    int valuesOffset = theBuffer.position() + (size << 2);
+
+    buffer.position(valuesOffset);
+    valueBuffers = Lists.newArrayList(buffer.slice());
+    buffer.position(indexOffset);
+    headerBuffer = buffer.slice();
+    final ByteBuffer valueBuffer = valueBuffers.get(0);
+    bufferIndexed = new BufferIndexed()
+    {
+      @Override
+      public T get(int index)
+      {
+        if (index < 0) {
+          throw new IAE("Index[%s] < 0", index);
+        }
+        if (index >= size) {
+          throw new IAE(String.format("Index[%s] >= size[%s]", index, size));
+        }
+
+        ByteBuffer copyHeaderBuffer = headerBuffer.asReadOnlyBuffer();
+        final int startOffset;
+        final int endOffset;
+
+        if (index == 0) {
+          copyHeaderBuffer.position(0);
+          startOffset = 4;
+          endOffset = copyHeaderBuffer.getInt();
+        } else {
+          copyHeaderBuffer.position((index - 1) * Ints.BYTES);
+          startOffset = copyHeaderBuffer.getInt() + 4;
+          endOffset = copyHeaderBuffer.getInt();
+        }
+        return _get(valueBuffer.asReadOnlyBuffer(), startOffset, endOffset);
+      }
+    };
+  }
+
+  GenericIndexed(
+      List<ByteBuffer> valueBuffs,
+      ByteBuffer headerBuff,
+      ObjectStrategy<T> strategy,
+      boolean allowReverseLookup,
+      int numElementsPer,
+      int numWritten
+  )
+  {
+    this.strategy = strategy;
+    this.allowReverseLookup = allowReverseLookup;
+    this.valueBuffers = valueBuffs;
+    this.headerBuffer = headerBuff;
+    this.size = numWritten;
+    this.numElementsPerValueFile = numElementsPer;
+    bufferIndexed = new BufferIndexed()
+    {
+      @Override
+      public T get(int index)
+      {
+        int fileNum = index / numElementsPerValueFile;
+        final ByteBuffer copyBuffer = valueBuffers.get(fileNum).asReadOnlyBuffer();
+        final ByteBuffer copyHeaderBuffer = headerBuffer.asReadOnlyBuffer();
```

Comment:
From the java doc I don't see any copying of data while creating ReadOnlyBuffer, it just ensures write protection on original buffer.